### ISSUE TITLE
Mixture returning NaN `log_prob` for valid sample (or: Softplus(Normal) returning `nan` probability on negative numbers)

### ISSUE THREAD

**Issue by agentydragon** [tensorflow/probability#1377]
I am working on some code for algorithmic trading. I'd like to create a probabilistic model over "take profit & stop loss" combinations. Those are valid if either:

* a) take profit is above current close, and stop loss is below current close,
* b) take profit is below current close, and stop loss is above current close.

(i.e., they cannot be on the same side).

This is a simplified version of my model, which makes a mixture distribution centered on `(0, 0)`:

```python
import tensorflow as tf
import tensorflow_probability as tfp
tfd = tfp.distributions
tfb = tfp.bijectors
dtype = tf.float64

long_d = tfd.MultivariateNormalDiag(loc=tf.zeros((1, 2,), dtype=dtype))
long_tp_sl = tfd.TransformedDistribution(
    long_d,
    tfb.Chain([
        tfb.Blockwise([
            tfb.Identity(),
            tfb.Scale(tf.constant(-1.0, dtype=dtype)),
        ],
                      block_sizes=[1, 1]),
        tfb.Softplus(),
    ]))

short_d = tfd.MultivariateNormalDiag(loc=tf.zeros((1, 2,), dtype=dtype))
short_tp_sl = tfd.TransformedDistribution(
    short_d,
    tfb.Chain([
        tfb.Blockwise([
            tfb.Scale(tf.constant(-1.0, dtype=dtype)),
            tfb.Identity(),
        ],
                      block_sizes=[1, 1]),
        tfb.Softplus(),
    ]))

trained_tp_sl = tfd.Mixture(
    cat=tfd.Categorical(probs=tf.constant([[0.5, 0.5]], dtype=dtype)),
    components=[long_tp_sl, short_tp_sl])

sample = trained_tp_sl.sample(500)
take_profit = sample[:, 0, 0]
stop_loss = sample[:, 0, 1]

plt.figure()
plt.scatter(x=take_profit, y=stop_loss, s=2)
```

This produces correct samples, as expected:

![image](https://user-images.githubusercontent.com/714892/125175558-668d1780-e1cd-11eb-808d-a7de49465c6c.png)

However, the `log_prob` of the returned sample is full of `nan`s:

```python3
>>> trained_tp_sl.log_prob(sample)
<tf.Tensor: shape=(500, 1), dtype=float64, numpy=
array([[nan],
       [nan],
       [nan],
...
```

I tried to reproduce this, and when I just take a mixture of 2 disjoint `Uniform` distributions, `prob` and `log_prob` seem to work fine:

```python3
import matplotlib.pyplot as plt
import tensorflow as tf
import tensorflow_probability as tfp

tfd = tfp.distributions

a = tfd.Uniform(low=-2.0, high=-1.0)
b = tfd.Uniform(low=1.0, high=2.0)
mix = tfd.Mixture(
    cat=tfd.Categorical(probs=[0.5, 0.5]),
    components=[a, b])

x = tf.linspace(-3.0, 3.0, num=100)

plt.figure()
plt.plot(x, a.prob(x), x, b.prob(x))

plt.figure()
plt.plot(x, mix.prob(x))

print(a.prob(1.5))  # -> 0.0
print(b.prob(1.5))  # -> 1.0
print(mix.prob(1.5))  # -> 0.5, OK
print()
print(a.log_prob(1.5))  # -> -inf
print(b.log_prob(1.5))  # -> 0
print(mix.log_prob(1.5))  # -> -0.6931472, OK
```

But I noticed that one my disjoint distributions return `nan` on samples where their probability is 0, instead of `-inf`:

```python3
>>> long_tp_sl.log_prob(sample)
<tf.Tensor: shape=(500, 1), dtype=float64, numpy=
array([[        nan],  # sample from short_tp_sl
       [        nan],
       [        nan],
       [        nan],
       [        nan],
       [-0.28825895],
       [-2.07584651],
       [        nan],
...
```

(On the other hand, when I tried this on the mix of disjoint Uniform distributions, notice the `log_prob` on samples with 0 probability was `-inf`.)

In fact, `Normal` distribution transformed with a `Softplus` returns `nan` probability on negative numbers:

```python3
>>> d = tfd.TransformedDistribution(tfd.Normal(loc=0.0, scale=1.0), tfb.Softplus())
>>> print(d.prob(-1.0))
tf.Tensor(nan, shape=(), dtype=float32)
```

The same also occurs with the `Exp` bijector.

Is there any way I can fix up my code to make the mixture's `log_prob` (and `prob`) behave correctly?

**Comment by axch** [comment]
That's a good one!

First, why does this happen, and what could TFP do about it?  The problem starts when `tfb.Softplus().inverse(-1.0) -> nan`.  That's arguably the right answer, though -- after all, for what `x` is it the case that `tfb.Softplus().forward(x) == -1.0`?  There are none.

From there, it's all downhill.  When `Normal` is asked for the log-prob of a `nan` sample, it has no choice: `Normal(...).log_prob(nan) -> nan`.  But, as you point out, the result is unfortunate: `dist = TransformedDistribution(Normal(...), Softplus())` is a perfectly well-defined distribution supported on positive reals, and you should be able to ask that `dist.log_prob(-1.0) == -inf`, not `nan`.

What can TFP do about this?  We could annotate all Bijectors with their support (on both sides), and add a flag to `TransformedDistribution` to return `-inf` log-probs on out-of-support queries.  (We would want this behavior to be flag-gated because we have a "pay for what you eat" performance philosophy, and we wouldn't want to push the cost of the check on users who know they will never have out-of-support queries.)

We could also try to approximate the above by adding a flag to `TransformedDistribution` to check whether the Bijector produced a `nan` from a non-`nan` input, and assume that means it was an out-of-support query.  That seems a little wonky, because `nan` can also mean "you're doing something mathematically reasonable but we can't compute the answer accurately", which should probably fail rather than silently producing a `-inf` log-prob.  But if it's flag-gated, it might be ok.

Could someone else from the TFP team weigh in on what would be a good solution on our end?

Now, what can you do about it?  You can monkey-patch `TransformedDistribution` to add the check I described above.  It might be cleaner to just write your own composite distribution, though.  Inherit from `TransformedDistribution` to get for free all the behavior that works correctly, and just override the broken bits.  Something like this:
```
class SoftplusInverseNormal(tfd.TransformedDistribution):

  def __init__(self, loc, scale):
    super(SoftplusInverseNormal, self).__init__(
      distribution=tfd.MultivariateNormalDiag(loc, scale),
      bijector=tfb.Softplus(),
    )

  def _log_prob(x):
    answer = super(SoftplusInverseNormal, self)._log_prob(x)
    return tf.where(x > 0, answer, -inf)
```

(and then further `TransformedDistribution` this to position it where it should be).  Check out our implementation of [LogNormal](https://github.com/tensorflow/probability/blob/main/tensorflow_probability/python/distributions/lognormal.py) for an example of how this is done.  One caveat: our naming convention for these derives from statistics convention: `LogNormal` means "log(x) ~ normal", not "x = log(y) where y ~ normal" as one might think.
